server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -161,6 +162,13 @@ public class IndexService extends AbstractIndexComponent implements IndicesClust
     private final ValuesSourceRegistry valuesSourceRegistry;
     private final Supplier<DocumentParsingObserver> documentParsingObserverSupplier;
 
+    // We keep track if a field has ever seen a value in an index at index level rather than
+    // on a per-shard basis. This has being done in order to reduce heap consumption on data
+    // nodes. The only limitation is that if a shard gets relocated to a different data node
+    // the response of that data node could be outdated but the _field_caps API response
+    // would still be correct, this works because we are not handling deletes.
+    private final Map<String, Boolean> fieldHasValue;
```

Comment:
Why use a map here now, the set we could get from `ConcurrentCollections` seems to me is a cleaner fit?

Reply:
Was done because of performance reason because putIfAbsent avoids the sync block. That was not the case before your change from last week (https://github.com/elastic/elasticsearch/pull/104704).
Changing it. :D Thanks